// @observablehq/framework/dist/config.js
import { createHash } from "node:crypto";
import { existsSync, readFileSync } from "node:fs";
import { stat } from "node:fs/promises";
import op from "node:path";
import { basename, dirname, join } from "node:path/posix";
import { cwd } from "node:process";
import { pathToFileURL } from "node:url";
import wrapAnsi from "wrap-ansi";
import { LoaderResolver } from "./dataloader.js";
import { visitMarkdownFiles } from "./files.js";
import { formatIsoDate, formatLocaleDate } from "./format.js";
import { createMarkdownIt, parseMarkdownMetadata } from "./markdown.js";
import { isAssetPath, parseRelativeUrl, resolvePath } from "./path.js";
import { resolveTheme } from "./theme.js";
import { bold, yellow } from "./tty.js";
function resolveConfig(configPath, root = ".") {
  return op.join(cwd(), root, configPath);
}
async function importConfig(path) {
  const { mtimeMs } = await stat(path);
  return (await import(`${pathToFileURL(path).href}?${mtimeMs}`)).default;
}
async function readConfig(configPath, root) {
  if (configPath === void 0)
    configPath = await resolveDefaultConfig(root);
  if (configPath === void 0)
    return normalizeConfig(void 0, root);
  return normalizeConfig(await importConfig(configPath), root, configPath);
}
async function resolveDefaultConfig(root) {
  const jsPath = resolveConfig("observablehq.config.js", root);
  if (existsSync(jsPath))
    return jsPath;
  const tsPath = resolveConfig("observablehq.config.ts", root);
  if (existsSync(tsPath))
    return await import("tsx/esm"), tsPath;
}
let cachedPages = null;
function readPages(root, md) {
  const files = [];
  const hash = createHash("sha256");
  for (const file of visitMarkdownFiles(root)) {
    if (file === "index.md" || file === "404.md")
      continue;
    const source = readFileSync(join(root, file), "utf8");
    files.push({ file, source });
    hash.update(file).update(source);
  }
  const key = hash.digest("hex");
  if (cachedPages?.key === key)
    return cachedPages.pages;
  const pages = [];
  for (const { file, source } of files) {
    const { data, title } = parseMarkdownMetadata(source, { path: file, md });
    if (data.draft)
      continue;
    const name = basename(file, ".md");
    const page = { path: join("/", dirname(file), name), name: title ?? "Untitled" };
    if (name === "index")
      pages.unshift(page);
    else
      pages.push(page);
  }
  cachedPages = { key, pages };
  return pages;
}
let currentDate = null;
function setCurrentDate(date) {
  currentDate = date;
}
const configCache = /* @__PURE__ */ new WeakMap();
function normalizeConfig(spec = {}, defaultRoot, watchPath) {
  const cachedConfig = configCache.get(spec);
  if (cachedConfig)
    return cachedConfig;
  const root = spec.root === void 0 ? findDefaultRoot(defaultRoot) : String(spec.root);
  const output = spec.output === void 0 ? "dist" : String(spec.output);
  const base = spec.base === void 0 ? "/" : normalizeBase(spec.base);
  const style = spec.style === null ? null : spec.style !== void 0 ? { path: String(spec.style) } : { theme: normalizeTheme(spec.theme === void 0 ? "default" : spec.theme) };
  const md = createMarkdownIt({
    linkify: spec.linkify === void 0 ? void 0 : Boolean(spec.linkify),
    typographer: spec.typographer === void 0 ? void 0 : Boolean(spec.typographer),
    quotes: spec.quotes === void 0 ? void 0 : spec.quotes,
    cleanUrls: spec.cleanUrls === void 0 ? void 0 : Boolean(spec.cleanUrls),
    markdownIt: spec.markdownIt
  });
  const title = spec.title === void 0 ? void 0 : String(spec.title);
  const pages = spec.pages === void 0 ? void 0 : normalizePages(spec.pages);
  const pager = spec.pager === void 0 ? true : Boolean(spec.pager);
  const toc = normalizeToc(spec.toc);
  const sidebar = spec.sidebar === void 0 ? void 0 : Boolean(spec.sidebar);
  const scripts = spec.scripts === void 0 ? [] : Array.from(spec.scripts, normalizeScript);
  const head = spec.head === void 0 ? "" : stringOrNull(spec.head);
  const header = spec.header === void 0 ? "" : stringOrNull(spec.header);
  const footer = spec.footer === void 0 ? defaultFooter() : stringOrNull(spec.footer);
  const search = Boolean(spec.search);
  const interpreters = normalizeInterpreters(spec.interpreters);
  const config = {
    root,
    output,
    base,
    title,
    sidebar,
    // see below
    pages,
    // see below
    pager,
    scripts,
    head,
    header,
    footer,
    toc,
    style,
    search,
    md,
    loaders: new LoaderResolver({ root, interpreters }),
    watchPath
  };
  if (pages === void 0)
    Object.defineProperty(config, "pages", { get: () => readPages(root, md) });
  if (sidebar === void 0)
    Object.defineProperty(config, "sidebar", { get: () => config.pages.length > 0 });
  configCache.set(spec, config);
  return config;
}
function defaultFooter() {
  const date = currentDate ?? /* @__PURE__ */ new Date();
  return `Built with <a href="https://observablehq.com/" target="_blank">Observable</a> on <a title="${formatIsoDate(
    date
  )}">${formatLocaleDate(date)}</a>.`;
}
function findDefaultRoot(defaultRoot) {
  if (defaultRoot !== void 0)
    return defaultRoot;
  const root = existsSync("docs") ? "docs" : "src";
  console.warn(
    wrapAnsi(
      `${yellow("Warning:")} the config file is missing the ${bold(
        "root"
      )} option, which specifies the path to the source root.${root === "docs" ? ` The recommended source root is ${bold('"src"')}; however, since ${bold(
        "docs"
      )} exists and was previously the default for this option, we will use ${bold('"docs"')}.` : ""} You can suppress this warning by specifying ${bold(`root: ${JSON.stringify(root)}`)} in the config file.
`,
      Math.min(80, process.stdout.columns ?? 80)
    )
  );
  return root;
}
function normalizeBase(spec) {
  let base = String(spec);
  if (!base.startsWith("/"))
    throw new Error(`base must start with slash: ${base}`);
  if (!base.endsWith("/"))
    base += "/";
  return base;
}
function normalizeTheme(spec) {
  return resolveTheme(typeof spec === "string" ? [spec] : spec === null ? [] : Array.from(spec, String));
}
function normalizeScript(spec) {
  const script = typeof spec === "string" ? { src: spec } : spec;
  const src = String(script.src);
  const async = script.async === void 0 ? false : Boolean(script.async);
  const type = script.type == null ? null : String(script.type);
  return { src, async, type };
}
function normalizePages(spec) {
  return Array.from(
    spec,
    (spec2) => "pages" in spec2 ? normalizeSection(spec2, (spec3) => normalizePage(spec3)) : normalizePage(spec2)
  );
}
function normalizeSection(spec, normalizePage2) {
  const name = String(spec.name);
  const collapsible = spec.collapsible === void 0 ? spec.open !== void 0 : Boolean(spec.collapsible);
  const open = collapsible ? Boolean(spec.open) : true;
  const pages = Array.from(spec.pages, normalizePage2);
  return { name, collapsible, open, pages };
}
function normalizePage(spec) {
  const name = String(spec.name);
  let path = String(spec.path);
  if (isAssetPath(path)) {
    const u = parseRelativeUrl(join("/", path));
    let { pathname } = u;
    pathname = pathname.replace(/\.html$/i, "");
    pathname = pathname.replace(/\/$/, "/index");
    path = pathname + u.search + u.hash;
  }
  return { name, path };
}
function normalizeInterpreters(spec = {}) {
  return Object.fromEntries(
    Object.entries(spec).map(([key, value]) => {
      return [String(key), value == null ? null : Array.from(value, String)];
    })
  );
}
function normalizeToc(spec = true) {
  const toc = typeof spec === "boolean" ? { show: spec } : spec;
  const label = toc.label === void 0 ? "Contents" : String(toc.label);
  const show = toc.show === void 0 ? true : Boolean(toc.show);
  return { label, show };
}
function mergeToc(spec = {}, toc) {
  const { label = toc.label, show = toc.show } = spec;
  return { label, show };
}
function mergeStyle(path, style, theme, defaultStyle) {
  return style === void 0 && theme === void 0 ? defaultStyle : style === null ? null : style !== void 0 ? { path: resolvePath(path, style) } : theme === void 0 ? defaultStyle : { theme };
}
function stringOrNull(spec) {
  return spec == null || spec === false ? null : String(spec);
}
export {
  mergeStyle,
  mergeToc,
  normalizeConfig,
  normalizeTheme,
  readConfig,
  setCurrentDate,
  stringOrNull
};
